File: tests/wiringPi/wiringSerial.py
import importlib
import pytest

m=importlib.import_module("wiringop.wiring_pi.wiring_serial")

def test_struct_serialconfig():
    obj=m.SerialConfig()
    _=obj.struct_size
    _=obj.version
    _=obj.baud
    _=obj.data_bits
    _=obj.parity
    _=obj.stop_bits
    _=obj.flow_control
    _=obj.xon_char
    _=obj.xoff_char
    _=obj.vmin
    _=obj.vtime
    assert len(obj.reserved) == 4

def test_struct_serialrs485config():
    obj=m.SerialRS485Config()
    _=obj.struct_size
    _=obj.flags
    _=obj.delay_before_send_ms
    _=obj.delay_after_send_ms
    _=obj.receive_address
    _=obj.destination_address
    assert len(obj.reserved8) == 2
    assert len(obj.reserved) == 4

def test_struct_serialcounters():
    obj=m.SerialCounters()
    _=obj.struct_size
    _=obj.cts
    _=obj.dsr
    _=obj.rng
    _=obj.dcd
    _=obj.rx
    _=obj.tx
    _=obj.frame
    _=obj.overrun
    _=obj.parity
    _=obj.brk
    _=obj.buf_overrun
    assert len(obj.reserved) == 4

def test_struct_serialhardwareinfo():
    obj=m.SerialHardwareInfo()
    _=obj.struct_size
    _=obj.hardware_type
    _=obj.fifo_size
    _=obj.baud_base
    _=obj.port_type
    _=obj.line
    _=obj.capabilities
    assert len(obj.reserved) == 6

def test_serial_open():
    _=m.serial_open("x", 1)

def test_serial_close():
    _=m.serial_close(1)

def test_serial_flush():
    _=m.serial_flush(1)

def test_serial_putchar():
    _=m.serial_putchar(1, 1)

def test_serial_puts():
    _=m.serial_puts(1, "x")

def test_serial_printf():
    m.serial_printf(1, "safe text")

def test_serial_data_avail():
    _=m.serial_data_avail(1)

def test_serial_getchar():
    _=m.serial_getchar(1)

def test_serial_config_init():
    c=m.serial_config_init();assert c.baud==9600 and c.data_bits==8

def test_serial_open_config():
    _=m.serial_open_config("x", m.SerialConfig())

def test_serial_get_config():
    rc,c=m.serial_get_config(7);assert rc==0

def test_serial_set_config():
    _=m.serial_set_config(1, m.SerialConfig())

def test_serial_get_baud():
    assert m.serial_get_baud(7)[0]==0

def test_serial_set_baud():
    _=m.serial_set_baud(1, 1)

def test_serial_read():
    assert m.serial_read(7,4)==b"0123"

def test_serial_write():
    assert m.serial_write(7,b"abcd")==4

def test_serial_read_timeout():
    assert m.serial_read_timeout(7,4,10,-1)==b"0123"

def test_serial_write_timeout():
    assert m.serial_write_timeout(7,b"abcd",10)==4

def test_serial_drain():
    _=m.serial_drain(1)

def test_serial_flush_input():
    _=m.serial_flush_input(1)

def test_serial_flush_output():
    _=m.serial_flush_output(1)

def test_serial_input_waiting():
    _=m.serial_input_waiting(1)

def test_serial_output_waiting():
    _=m.serial_output_waiting(1)

def test_serial_send_break():
    _=m.serial_send_break(1, 1)

def test_serial_set_break():
    _=m.serial_set_break(1, 1)

def test_serial_get_modem_lines():
    assert m.serial_get_modem_lines(7)[0]==0

def test_serial_set_modem_lines():
    _=m.serial_set_modem_lines(1, 1, 1)

def test_serial_set_rts():
    _=m.serial_set_rts(1, 1)

def test_serial_set_dtr():
    _=m.serial_set_dtr(1, 1)

def test_serial_get_cts():
    _=m.serial_get_cts(1)

def test_serial_get_dsr():
    _=m.serial_get_dsr(1)

def test_serial_get_ri():
    _=m.serial_get_ri(1)

def test_serial_get_cd():
    _=m.serial_get_cd(1)

def test_serial_set_input_flow():
    _=m.serial_set_input_flow(1, 1)

def test_serial_set_output_flow():
    _=m.serial_set_output_flow(1, 1)

def test_serial_set_exclusive():
    _=m.serial_set_exclusive(1, 1)

def test_serial_get_low_latency():
    _=m.serial_get_low_latency(1)

def test_serial_set_low_latency():
    _=m.serial_set_low_latency(1, 1)

def test_serial_get_counters():
    assert m.serial_get_counters(7)[0]==0

def test_serial_tx_empty():
    _=m.serial_tx_empty(1)

def test_serial_get_rs485():
    assert m.serial_get_rs485(7)[0]==0

def test_serial_set_rs485():
    _=m.serial_set_rs485(1, m.SerialRS485Config())

def test_serial_get_rx_trigger():
    _=m.serial_get_rx_trigger(1)

def test_serial_set_rx_trigger():
    _=m.serial_set_rx_trigger(1, 1)

def test_serial_get_wakeup():
    _=m.serial_get_wakeup(1)

def test_serial_set_wakeup():
    _=m.serial_set_wakeup(1, 1)

def test_serial_get_hardware_info():
    assert m.serial_get_hardware_info(7)[0]==0

def test_serial_get_capabilities():
    assert m.serial_get_capabilities(7)[0]==0

def test_serial_context_open():
    cfg=m.serial_config_init();a=m.serial_context_open("mock",cfg);assert isinstance(a,int) and a!=0;m.serial_context_close(a)

def test_serial_context_from_fd():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_get_fd(a)==7;m.serial_context_close(a)

def test_serial_context_close():
    cfg=m.serial_config_init();a=m.serial_context_open("mock",cfg);m.serial_context_close(a)

def test_serial_context_get_fd():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_get_fd(a)==7;m.serial_context_close(a)

def test_serial_context_read():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_read(a,4,10,-1)==b"0123";m.serial_context_close(a)

def test_serial_context_write():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_write(a,b"abcd",10)==4;m.serial_context_close(a)

def test_serial_context_cancel_read():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_cancel_read(a)==0;m.serial_context_close(a)

def test_serial_context_cancel_write():
    a=m.serial_context_from_fd(7,0);assert m.serial_context_cancel_write(a)==0;m.serial_context_close(a)

def test_constants_and_macros():
    assert hasattr(m,"SERIAL_CONFIG_VERSION")
    assert hasattr(m,"SERIAL_CONFIG_STRUCT_SIZE")
    assert hasattr(m,"SERIAL_PARITY_NONE")
    assert hasattr(m,"SERIAL_PARITY_ODD")
    assert hasattr(m,"SERIAL_PARITY_EVEN")
    assert hasattr(m,"SERIAL_PARITY_MARK")
    assert hasattr(m,"SERIAL_PARITY_SPACE")
    assert hasattr(m,"SERIAL_STOP_BITS_ONE")
    assert hasattr(m,"SERIAL_STOP_BITS_ONE_POINT_FIVE")
    assert hasattr(m,"SERIAL_STOP_BITS_TWO")
    assert hasattr(m,"SERIAL_FLOW_NONE")
    assert hasattr(m,"SERIAL_FLOW_XON_XOFF")
    assert hasattr(m,"SERIAL_FLOW_RTS_CTS")
    assert hasattr(m,"SERIAL_MODEM_RTS")
    assert hasattr(m,"SERIAL_MODEM_DTR")
    assert hasattr(m,"SERIAL_MODEM_CTS")
    assert hasattr(m,"SERIAL_MODEM_DSR")
    assert hasattr(m,"SERIAL_MODEM_RI")
    assert hasattr(m,"SERIAL_MODEM_CD")
    assert hasattr(m,"SERIAL_MODEM_LOOP")
    assert hasattr(m,"SERIAL_RS485_ENABLED")
    assert hasattr(m,"SERIAL_RS485_RTS_ON_SEND")
    assert hasattr(m,"SERIAL_RS485_RTS_AFTER_SEND")
    assert hasattr(m,"SERIAL_RS485_RX_DURING_TX")
    assert hasattr(m,"SERIAL_RS485_TERMINATE_BUS")
    assert hasattr(m,"SERIAL_RS485_ADDRESS_MODE")
    assert hasattr(m,"SERIAL_RS485_RX_ADDRESS")
    assert hasattr(m,"SERIAL_RS485_DEST_ADDRESS")
    assert hasattr(m,"SERIAL_RS485_MODE_RS422")
    assert hasattr(m,"SERIAL_CAP_CUSTOM_BAUD")
    assert hasattr(m,"SERIAL_CAP_MARK_SPACE_PARITY")
    assert hasattr(m,"SERIAL_CAP_XON_XOFF")
    assert hasattr(m,"SERIAL_CAP_RTS_CTS")
    assert hasattr(m,"SERIAL_CAP_MODEM_CONTROL")
    assert hasattr(m,"SERIAL_CAP_MODEM_STATUS")
    assert hasattr(m,"SERIAL_CAP_EXCLUSIVE")
    assert hasattr(m,"SERIAL_CAP_LOW_LATENCY")
    assert hasattr(m,"SERIAL_CAP_RS485")
    assert hasattr(m,"SERIAL_CAP_ICOUNT")
    assert hasattr(m,"SERIAL_CAP_TX_EMPTY")
    assert hasattr(m,"SERIAL_CAP_RX_TRIGGER")
    assert hasattr(m,"SERIAL_CAP_WAKEUP")
    assert hasattr(m,"SERIAL_CAP_8250")
    assert hasattr(m,"SERIAL_CAP_DW_APB_UART")
    assert hasattr(m,"SERIAL_CAP_RK3588_UART")
    assert hasattr(m,"SERIAL_CAP_DMA_DESCRIBED")
    assert hasattr(m,"SERIAL_CAP_AUTO_RTS_CTS")
    assert hasattr(m,"SERIAL_CAP_FIFO")
    assert hasattr(m,"SERIAL_CAP_CANCEL_IO")
    assert hasattr(m,"SERIAL_HARDWARE_UNKNOWN")
    assert hasattr(m,"SERIAL_HARDWARE_8250")
    assert hasattr(m,"SERIAL_HARDWARE_DW_APB_UART")
    assert hasattr(m,"SERIAL_HARDWARE_RK3588_UART")

def test_serial_port_high_level():
    p=m.SerialPort("mock",115200)
    assert p.is_open and p.fileno()==7 and p.port=="mock" and p.name=="mock"
    p.baudrate=123456;assert p.baudrate==123456 and p.baud==123456
    p.bytesize=8;p.parity="N";p.stopbits=1;p.xonxoff=True;p.rtscts=True
    p.timeout=0.01;p.write_timeout=0.01;p.inter_byte_timeout=None
    assert p.read(4)==b"0123"
    b=bytearray(4);assert p.read_into(b)==4 and b==b"0123"
    assert p.write(memoryview(b"abcd"))==4
    assert p.in_waiting==3 and p.out_waiting==3
    assert p.hardware_info.fifo_size==64
    assert p.capabilities & m.SERIAL_CAP_RK3588_UART
    assert p.capabilities & m.SERIAL_CAP_CANCEL_IO
    assert p.cts and p.dsr and p.ri and p.cd
    p.rts=True;p.dtr=True;p.set_input_flow_control(True);p.set_output_flow_control(True)
    p.break_condition=True;assert p.break_condition;p.break_condition=False
    p.cancel_read();p.cancel_write()
    p.close();assert not p.is_open
    p.open();assert p.is_open;p.close()

def test_serial_port_dsrdtr_reports_linux_limit():
    p=m.SerialPort("mock",9600)
    assert p.dsrdtr is False
    with pytest.raises(NotImplementedError): p.dsrdtr=True
    p.close()

File: support/mock_backend/wiringPi/wiringSerial.c
#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>
#include <string.h>
#include "wiringPi/wiringSerial.h"
struct serialContext { int fd; int ownFd; int cancelRead[2]; int cancelWrite[2]; };
static serialConfig mock_serial_config={.structSize=sizeof(serialConfig),.version=SERIAL_CONFIG_VERSION,.baud=9600,.dataBits=8,.stopBits=SERIAL_STOP_BITS_ONE};
static serialRS485Config mock_rs485={.structSize=sizeof(serialRS485Config)};
static struct serialContext mock_context={.fd=7};
static int mock_rx_trigger=16;static int mock_wakeup=0;
int serialOpen(const char * device, const int baud) { (void)device;mock_serial_config.baud=(uint32_t)baud;return 7; }
void serialClose(const int fd) { (void)fd; }
void serialFlush(const int fd) { (void)fd; }
void serialPutchar(const int fd, const unsigned char c) { (void)fd;(void)c; }
void serialPuts(const int fd, const char * s) { (void)fd;(void)s; }
void serialPrintf(const int fd, const char * message, ...) { (void)fd;(void)message; }
int serialDataAvail(const int fd) { (void)fd;return 1; }
int serialGetchar(const int fd) { (void)fd;return 1; }
void serialConfigInit(serialConfig * config) { memset(config,0,sizeof(*config));config->structSize=sizeof(*config);config->version=SERIAL_CONFIG_VERSION;config->baud=9600;config->dataBits=8;config->parity=SERIAL_PARITY_NONE;config->stopBits=SERIAL_STOP_BITS_ONE;config->xonChar=0x11;config->xoffChar=0x13; }
int serialOpenConfig(const char * device, const serialConfig * config) { (void)device;if(config)mock_serial_config=*config;return 7; }
int serialGetConfig(const int fd, serialConfig * config) { (void)fd;*config=mock_serial_config;return 0; }
int serialSetConfig(const int fd, const serialConfig * config) { (void)fd;mock_serial_config=*config;return 0; }
int serialGetBaud(const int fd, unsigned int * baud) { (void)fd;*baud=mock_serial_config.baud;return 0; }
int serialSetBaud(const int fd, const unsigned int baud) { (void)fd;mock_serial_config.baud=baud;return 0; }
ssize_t serialRead(const int fd, void * buffer, const size_t count) { (void)fd;for(size_t i=0;i<count;++i)((unsigned char*)buffer)[i]=(unsigned char)(0x30+(i%10));return(ssize_t)count; }
ssize_t serialWrite(const int fd, const void * buffer, const size_t count) { (void)fd;(void)buffer;return(ssize_t)count; }
ssize_t serialReadTimeout(const int fd, void * buffer, const size_t count, const int timeoutMs, const int interByteTimeoutMs) { (void)fd;(void)timeoutMs;(void)interByteTimeoutMs;for(size_t i=0;i<count;++i)((unsigned char*)buffer)[i]=(unsigned char)(0x30+(i%10));return(ssize_t)count; }
ssize_t serialWriteTimeout(const int fd, const void * buffer, const size_t count, const int timeoutMs) { (void)fd;(void)buffer;(void)timeoutMs;return(ssize_t)count; }
int serialDrain(const int fd) { (void)fd;return 1; }
int serialFlushInput(const int fd) { (void)fd;return 1; }
int serialFlushOutput(const int fd) { (void)fd;return 1; }
int serialInputWaiting(const int fd) { (void)fd;return 3; }
int serialOutputWaiting(const int fd) { (void)fd;return 3; }
int serialSendBreak(const int fd, const unsigned int durationMs) { (void)fd;(void)durationMs;return 1; }
int serialSetBreak(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialGetModemLines(const int fd, unsigned int * lines) { (void)fd;*lines=SERIAL_MODEM_RTS|SERIAL_MODEM_DTR|SERIAL_MODEM_CTS;return 0; }
int serialSetModemLines(const int fd, const unsigned int setMask, const unsigned int clearMask) { (void)fd;(void)setMask;(void)clearMask;return 1; }
int serialSetRTS(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialSetDTR(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialGetCTS(const int fd) { (void)fd;return 1; }
int serialGetDSR(const int fd) { (void)fd;return 1; }
int serialGetRI(const int fd) { (void)fd;return 1; }
int serialGetCD(const int fd) { (void)fd;return 1; }
int serialSetInputFlow(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialSetOutputFlow(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialSetExclusive(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialGetLowLatency(const int fd) { (void)fd;return 1; }
int serialSetLowLatency(const int fd, const int enabled) { (void)fd;(void)enabled;return 1; }
int serialGetCounters(const int fd, serialCounters * counters) { (void)fd;memset(counters,0,sizeof(*counters));counters->structSize=sizeof(*counters);counters->rx=12;counters->tx=34;return 0; }
int serialTxEmpty(const int fd) { (void)fd;return 1; }
int serialGetRS485(const int fd, serialRS485Config * config) { (void)fd;*config=mock_rs485;config->structSize=sizeof(*config);return 0; }
int serialSetRS485(const int fd, const serialRS485Config * config) { (void)fd;mock_rs485=*config;return 0; }
int serialGetRxTrigger(const int fd) { (void)fd;return mock_rx_trigger; }
int serialSetRxTrigger(const int fd, const unsigned int bytes) { (void)fd;mock_rx_trigger=(int)bytes;return 0; }
int serialGetWakeup(const int fd) { (void)fd;return mock_wakeup; }
int serialSetWakeup(const int fd, const int enabled) { (void)fd;mock_wakeup=enabled;return 0; }
int serialGetHardwareInfo(const int fd, serialHardwareInfo * info) { (void)fd;memset(info,0,sizeof(*info));info->structSize=sizeof(*info);info->hardwareType=SERIAL_HARDWARE_RK3588_UART;info->fifoSize=64;info->capabilities=SERIAL_CAP_CUSTOM_BAUD|SERIAL_CAP_RTS_CTS|SERIAL_CAP_RK3588_UART|SERIAL_CAP_FIFO|SERIAL_CAP_CANCEL_IO;return 0; }
int serialGetCapabilities(const int fd, uint64_t * capabilities) { (void)fd;*capabilities=SERIAL_CAP_CUSTOM_BAUD|SERIAL_CAP_RTS_CTS|SERIAL_CAP_RK3588_UART|SERIAL_CAP_FIFO|SERIAL_CAP_CANCEL_IO;return 0; }
serialContext * serialContextOpen(const char * device, const serialConfig * config) { (void)device;if(config)mock_serial_config=*config;mock_context.fd=7;return &mock_context; }
serialContext * serialContextFromFd(const int fd, const int takeOwnership) { (void)takeOwnership;mock_context.fd=fd;return &mock_context; }
void serialContextClose(serialContext * context) { (void)context; }
int serialContextGetFd(const serialContext * context) { return context?context->fd:-1; }
ssize_t serialContextRead(serialContext * context, void * buffer, const size_t count, const int timeoutMs, const int interByteTimeoutMs) { (void)timeoutMs;(void)interByteTimeoutMs;if(!context)return-1;for(size_t i=0;i<count;++i)((unsigned char*)buffer)[i]=(unsigned char)(0x30+(i%10));return(ssize_t)count; }
ssize_t serialContextWrite(serialContext * context, const void * buffer, const size_t count, const int timeoutMs) { (void)timeoutMs;if(!context)return-1;(void)buffer;return(ssize_t)count; }
int serialContextCancelRead(serialContext * context) { return context?0:-1; }
int serialContextCancelWrite(serialContext * context) { return context?0:-1; }
